File: shared/types/report.ts
export type ReportFormat = "pdf" | "markdown" | "html" | "both" | "all";

/** Formats that yield a downloadable artifact (one URL per format). */
export type ReportDownloadFormat = "pdf" | "markdown" | "html";
export type ReportStatus = "pending" | "generating" | "completed" | "failed";
export type ReportPeriod = "24h" | "7d" | "30d" | "90d";
export type ReportScheduleCadence = "weekly" | "monthly";
export type ReportScheduleRunStatus =
  | "pending"
  | "generating"
  | "delivering"
  | "completed"
  | "failed";
export type ReportScheduleDeliveryChannel = "email" | "slack";

export interface ReportCreateParams {
  scanId: string;
  monitorId?: string;
  monitorPeriod?: ReportPeriod;
  format?: ReportFormat;
  title?: string;
}

export interface ReportRecord {
  id: string;
  title: string;
  format: ReportFormat;
  status: ReportStatus;
  scanId: string | null;
  monitorId: string | null;
  monitorPeriod: ReportPeriod | null;
  fileSizeBytes: number | null;
  errorMessage: string | null;
  reportMeta?: Record<string, unknown> | null;
  createdAt: string;
  completedAt: string | null;
}

export interface ReportListItem {
  id: string;
  title: string;
  format: ReportFormat;
  status: ReportStatus;
  /**
   * Backing scan id, mirrored from ``Report.scan_id``. May be null when the
   * source scan has been deleted (FK is ON DELETE SET NULL). Used by the
   * Reports list page to deep-link into the scan-to-scan diff (T5.2).
   */
  scanId: string | null;
  scanDomain: string | null;
  fileSizeBytes: number | null;
  createdAt: string;
  completedAt: string | null;
}

export interface ReportListResult {
  reports: ReportListItem[];
  meta?: {
    page: number;
    limit: number;
    total: number;
    status?: ReportStatus | null;
  };
}

export interface ReportPreview {
  id: string;
  title: string;
  status: ReportStatus;
  contentMd: string;
  reportMeta?: Record<string, unknown> | null;
}

export interface ReportScheduleRun {
  id: string;
  scheduleId: string;
  reportId: string | null;
  status: ReportScheduleRunStatus;
  startedAt: string | null;
  completedAt: string | null;
  errorMessage: string | null;
  deliverySummary: Record<string, unknown> | null;
}

export interface ReportSchedule {
  id: string;
  userId: number;
  name: string;
  scanId: string | null;
  monitorId: string | null;
  monitorPeriod: ReportPeriod | null;
  format: ReportFormat;
  cadence: ReportScheduleCadence;
  timezone: string;
  dayOfWeek: number | null;
  dayOfMonth: number | null;
  hour: number;
  minute: number;
  deliveryChannels: ReportScheduleDeliveryChannel[];
  emailRecipients: string[];
  isEnabled: boolean;
  lastRunAt: string | null;
  nextRunAt: string | null;
  createdAt: string;
  updatedAt: string;
  recentRuns: ReportScheduleRun[];
}

export interface ReportScheduleCreateParams {
  name: string;
  scanId: string;
  monitorId?: string | null;
  monitorPeriod?: ReportPeriod;
  format?: ReportFormat;
  cadence: ReportScheduleCadence;
  timezone: string;
  dayOfWeek?: number | null;
  dayOfMonth?: number | null;
  hour: number;
  minute: number;
  deliveryChannels: ReportScheduleDeliveryChannel[];
  emailRecipients: string[];
  isEnabled?: boolean;
}

export type ReportScheduleUpdateParams = Partial<ReportScheduleCreateParams>;

export interface ReportScheduleListResult {
  schedules: ReportSchedule[];
}

export interface ReportScheduleRunsResult {
  runs: ReportScheduleRun[];
}

File: backend/app/api/v1/schemas/report.py
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from app.models.report import ReportFormat, ReportStatus

REPORT_PERIODS = ("24h", "7d", "30d", "90d")


class ReportCreateRequest(BaseModel):
    scan_id: uuid.UUID
    monitor_id: uuid.UUID | None = None
    monitor_period: Literal["24h", "7d", "30d", "90d"] = "30d"
    format: Literal["pdf", "markdown", "html", "both", "all"] = "pdf"
    title: str | None = Field(default=None, max_length=512)

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class ReportResponse(BaseModel):
    id: uuid.UUID
    title: str
    format: ReportFormat
    status: ReportStatus
    scan_id: uuid.UUID | None = None
    monitor_id: uuid.UUID | None = None
    monitor_period: str | None = None
    file_size_bytes: int | None = None
    error_message: str | None = None
    report_meta: dict[str, Any] | None = None
    created_at: datetime
    completed_at: datetime | None = None

    model_config = ConfigDict(
        from_attributes=True,
        alias_generator=to_camel,
        populate_by_name=True,
    )


class ReportListItem(BaseModel):
    id: uuid.UUID
    title: str
    format: ReportFormat
    status: ReportStatus
    # Scan id is included so the dashboard can deep-link from a report row
    # into the scan-to-scan diff page (Phase 5 / T5.2). May be null when the
    # underlying scan has been deleted (FK is set to NULL on cascade).
    scan_id: uuid.UUID | None = None
    scan_domain: str | None = None
    file_size_bytes: int | None = None
    created_at: datetime
    completed_at: datetime | None = None

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class ReportPreviewResponse(BaseModel):
    id: uuid.UUID
    title: str
    status: ReportStatus
    content_md: str
    report_meta: dict[str, Any] | None = None

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class ReportListResponse(BaseModel):
    reports: list[ReportListItem]

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)
